src/mpl/configure.ac

```diff
@@ -197,30 +208,101 @@ if test "$enable_fast_alwaysinline" = "yes"; then
-                               pac_cv_found_sse2_runnable="unknown")
-                ])
-PAC_POP_FLAG([CFLAGS])
-if test "$pac_cv_found_sse2" = "yes" && test "$pac_cv_found_sse2_runnable" = "yes"; then
+if test "$force_enable_sse2" = "yes"; then
     PAC_APPEND_FLAG([-msse2],[CFLAGS])
-    AC_CACHE_CHECK([whether _mm_stream_si128 is supported], pac_cv_found__mm_stream_si128,[
+fi
+
+AC_CACHE_CHECK([whether -mavx is supported], pac_cv_found_avx,
+              [PAC_C_CHECK_COMPILER_OPTION([-mavx],pac_cv_found_avx=yes,pac_cv_found_avx=no)],
+              pac_cv_found_avx=no,pac_cv_found_avx=yes)
+if test "$force_enable_avx" = "yes"; then
+    PAC_APPEND_FLAG([-mavx],[CFLAGS])
+fi
+
+AC_CACHE_CHECK([whether -mavx512f is supported], pac_cv_found_avx512f,
+               [PAC_C_CHECK_COMPILER_OPTION([-mavx512f],pac_cv_found_avx512f=yes,pac_cv_found_avx512f=no)],
+               pac_cv_found_avx512f=no,pac_cv_found_avx512f=yes)
+
+if test "$force_enable_avx512f" = "yes"; then
+    PAC_APPEND_FLAG([-mavx512f],[CFLAGS])
+fi
+
+if test "$enable_fast_instr_auto" = "yes"; then
+    PAC_PUSH_FLAG([CFLAGS])
+    PAC_APPEND_FLAG([-mavx512f],[CFLAGS])
+    AC_CACHE_CHECK([whether AVX512F is supported by the CPU], pac_cv_found_avx512f_runnable,[
                     AC_RUN_IFELSE([AC_LANG_SOURCE([[
                                    #include <immintrin.h>
+
+                                   int main() {
+                                       __m512i a = _mm512_set1_epi32(1);
+                                       __asm__ volatile("" : : "x" (a) : "memory");
+                                       return 0;
+                                   }
+                                   ]])], pac_cv_found_avx512f_runnable="yes",
+                                   pac_cv_found_avx512f_runnable="no",
+                                   pac_cv_found_avx512f_runnable="unknown")
+                    ])
+    PAC_POP_FLAG([CFLAGS])
+    if test "$pac_cv_found_avx512f_runnable" = "yes"; then
+        PAC_APPEND_FLAG([-mavx512f],[CFLAGS])
+    else
+        # if avx512f, try avx
+        PAC_PUSH_FLAG([CFLAGS])
+        PAC_APPEND_FLAG([-mavx],[CFLAGS])
+        AC_CACHE_CHECK([whether AVX is supported by the CPU], pac_cv_found_avx_runnable,[
+                        AC_RUN_IFELSE([AC_LANG_SOURCE([[
+                                       #include <immintrin.h>
+
+                                       int main() {
+                                           __m256i a = _mm256_set1_epi32(1);
+                                           __asm__ volatile("" : : "x" (a) : "memory");
+                                           return 0;
+                                       }
+                                       ]])], pac_cv_found_avx_runnable="yes",
+                                       pac_cv_found_avx_runnable="no",
+                                       pac_cv_found_avx_runnable="unknown")
+                        ])
+        PAC_POP_FLAG([CFLAGS])
```

Comment:
why do we do these 2-stage checks where we try to compile a simple program then a more complicated one? if success means building the mroe complicated one, then just do that. also since the program source is repeated multiple times, we should store it in a shell variable and reuse it.